lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -308,19 +308,20 @@ static LogicalResult verifyComponentOp(ComponentOp op) {
 }
 
 void ComponentOp::build(OpBuilder &builder, OperationState &result,
-                        StringAttr name, ArrayRef<ComponentPortInfo> ports) {
+                        StringAttr name, ArrayRef<ComponentPortInfo> inputs,
```

Comment:
I see two alternatives to this.

1. Separately track port directions as an attribute, similar to [FIRRTL](https://github.com/llvm/circt/blob/ee457605169c16b1ce1400201e66d06d723f282b/include/circt/Dialect/FIRRTL/FIRRTLOps.h#L40-L53). We would no longer need this pre-condition that inputs are listed before outputs.

2.  Keeping this as a single list of `ports`, and then partitioning the ports by direction `{inputs, outputs}`. 

Something like:
```
 auto portIt = llvm::partition(ports, [](auto port){ return port.direction == INPUT; });
 uint64_t numInPorts = std::distance(ports.begin(), portIt);
 ...
```

Reasoning: The current changes redundantly lists input and output ports, which should already be identified by the `ComponentPortInfo::PortDirection`.

WDYT?

Reply:
Regarding option 2, if the partition is deterministic in terms of not reordering the inputs and the outputs themselves, then that would work. But option 1 would get closer to solving the actual issue, being that port directionality throughout the ComponentOp functions is being tracked as information based on the provided arguments during construction and the deduced # of inputs.

Reply:
5b646b9 implements option 1.